### PR TITLE
Fix markdown anchor links

### LINKED ISSUE #9672: Markdown anchor links don't work
- Gitea version (or commit ref): a311ace354c4a00e0576943b0b0bff651fbcb066 ([master](https://github.com/go-gitea/gitea/tree/master))
- Git version: 2.24.1
- Operating system: Arch Linux
- Database: SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - Yes, https://try.gitea.io/adfjiowfeiofhaehioh/test-markdown/src/branch/master/test.md

## Description

Anchor links (to headers) in markdown link to `#anchor` instead of `#user-content-anchor` (modified).
See the [provided link](https://try.gitea.io/adfjiowfeiofhaehioh/test-markdown/src/branch/master/test.md).

### PR DESCRIPTION
**Pull request by mohe2015**
Fix https://github.com/go-gitea/gitea/issues/9672

This fixes markdown anchor links by prepending `#user-content` to it.

If this is not the fix you are looking or if you need tests feel free to create your own pull request.
Unfortunately I don't have the time to do that.

### CODE REVIEW

File: modules/markup/markdown/goldmark.go [hunk 1]
```diff
@@ -79,6 +79,9 @@ func (g *GiteaASTTransformer) Transform(node *ast.Document, reader text.Reader,
 				}
 				link = []byte(giteautil.URLJoin(pc.Get(urlPrefixKey).(string), lnk))
 			}
+			if len(link) > 0 && link[0] == '#' {
```
Comment (guillep2k) [inline modules/markup/markdown/goldmark.go:82]: The `if` can be rearranged with the code above so the conditions are only checked once (e.g. `len(link)>0` and `link[0]=='#'`)

**Comment by zeripath** [comment]
@jolheiser What do you think of this. I remember us discussing doing this when we still had blackfriday. Was it much harder to do this fix on the blackfriday side?

**Comment by jolheiser** [comment]
> @jolheiser What do you think of this. I remember us discussing doing this when we still had blackfriday. Was it much harder to do this fix on the blackfriday side?

I don't remember exactly, but I think this would have broken other things in blackfriday that would have required more fixes, so we decided it wasn't worth it (at the time).

File: modules/markup/markdown/goldmark.go (hunk 1 above)
Comment (mohe2015) [reply modules/markup/markdown/goldmark.go:82]: I don't know if this is more readable so I would like some other opinions whether this is worth it.

Comment (techknowlogick) [reply modules/markup/markdown/goldmark.go:82]: @guillep2k the first `if` has `link[0] != '#'` and the second has `link[0] == '#'` (note: the `!` in the first that isn't in second), so I think that even though there is one conditional in the `if` that overlaps, any computation time that would be saved is minimal, and potentially even make the whole section of code be less readable (but that is from my rough attempts to combine them which I haven't spent too much time trying, I'd be happy to be proven wrong).

**Review by zeripath — COMMENTED** [review]
We probably shouldn't add user-content- to links that already have it. Although I dunno.

I'm also wondering about whether we'd ever want to let markdown rendered relative links link to non rendered content - is there a way of doing that? I guess those should be fully qualified as we can't tell when markdown is rendered.

**Comment by zeripath** [comment]
Hmm I think I'll approve this. 

If we do merge it we should backport it to 1.11

**Review by zeripath — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by guillep2k** [comment]
Sorry I'm late, but how can we link to another comment in a thread with this change? 😕

**Comment by zeripath** [comment]
You'd need to use the fully qualified link for it. (Which is what you should do anyway as you can't predict when the markdown is being rendered that you're on the comment page.)

**Comment by zeripath** [comment]
e.g. https://github.com/go-gitea/gitea/pull/9673#issuecomment-575176485

or [/go-gitea/gitea/pull/9673#issuecomment-575176485](https://github.com/go-gitea/gitea/pull/9673#issuecomment-575176485)

**Comment by guillep2k** [comment]
> 
> 
> You'd need to use the fully qualified link for it. (Which is what you should do anyway as you can't predict when the markdown is being rendered that you're on the comment page.)

Hummm..... should we write a parser for our migration to `gitea.com`, replacing `https://github.com/go-gitea/gitea/` with `https://gitea.com/gitea/gitea/`?

**Comment by lafriks** [comment]
This could be easily done with sql update with replace

**Comment by sapk** [comment]
@mohe2015 can you send a backport ?

**Pull request merged**